src/connector/src/parser/additional_columns.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::catalog::{ColumnCatalog, ColumnDesc, ColumnId};
+use risingwave_common::types::DataType;
+use risingwave_pb::plan_common::AdditionalColumnType;
+
+use crate::source::{
+    KAFKA_CONNECTOR, KINESIS_CONNECTOR, OPENDAL_S3_CONNECTOR, PULSAR_CONNECTOR, S3_CONNECTOR,
+};
+
+pub type CompatibleAdditionalColumnsFn =
+    Box<dyn Fn(ColumnId, &str) -> ColumnCatalog + Send + Sync + 'static>;
+
+pub fn get_connector_compatible_additional_columns(
+    connector_name: &str,
+) -> Option<Vec<(&'static str, CompatibleAdditionalColumnsFn)>> {
```

Comment:
(Continue from https://github.com/risingwavelabs/risingwave/pull/13707#discussion_r1427561433)

I think it's not necessary to return a function here (aka. high-order function). Exemplified in https://github.com/risingwavelabs/risingwave/pull/13707#discussion_r1427561433:

```rust
pub fn get_additional_column_catalog(connector_name: &str, column_name: &str, column_id: &ColumnId) -> Result<ColumnCatalog>
```

Reply:
Oh sorry, I misunderstood your comments. 

```
pub fn get_additional_column_catalog(connector_name: &str, column_name: &str, column_id: &ColumnId) -> Result<ColumnCatalog>
```

It is not an acceptable solution for me because I intend to insert all accepted columns in order.
For example, the output schema will always follow the same order `(key, partition, offset)`

```
1)
include key as aaa, partition as bbb, offset as ccc
2)
include offset as ccc, partition as bbb, key as aaa
```